### PR TITLE
Forbidden character in username injection

### PR DESCRIPTION
**Pull request by chibenwa**
Extracted and reworked from @mbaechler  PR comments on https://github.com/linagora/james-project/pull/1087

### CODE REVIEW

**Comment by chibenwa** [comment]
test this please

Output of the test is lost.

**Comment by aduprat** [comment]
```
[INFO] Apache James :: Server :: Data :: Library .......... FAILURE [  4.326 s]

[ERROR]   VirtualHostingNameValidatorTest.validateShouldThrowOnMissingLocalPartWhenVirtualHosting:65 
Expecting:
  <java.lang.IllegalArgumentException: username should not be empty>
to be an instance of:
  <org.apache.james.user.api.UsersRepositoryException>
but was:
  <"java.lang.IllegalArgumentException: username should not be empty
```

File: server/data/data-library/src/main/java/org/apache/james/user/lib/UsernameValidatorAggregator.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.Set;
+
+import javax.inject.Inject;
+
+import org.apache.james.user.api.UsersRepositoryException;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.fge.lambdas.consumers.ThrowingConsumer;
+
+public class UsernameValidatorAggregator implements UsernameValidator {
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/user/lib/UsernameValidatorAggregator.java]: It's probably more a Composite or a Chain than an aggregator (aggregator is often related to output)

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/data/MemoryDataModule.java [hunk 2]
```diff
@@ -55,6 +57,11 @@ protected void configure() {
         bind(MemoryUsersRepository.class).toInstance(MemoryUsersRepository.withVirtualHosting());
         bind(UsersRepository.class).to(MemoryUsersRepository.class);
 
+
```
Comment (aduprat) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/data/MemoryDataModule.java]: extra lines

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/data/CassandraUsersRepositoryModule.java [hunk 3]
```diff
@@ -39,6 +41,9 @@
 public class CassandraUsersRepositoryModule extends AbstractModule {
     @Override
     public void configure() {
+        bind(UsernameValidatorAggregator.class).in(Scopes.SINGLETON);
+        bind(UsernameValidator.class).to(UsernameValidatorAggregator.class);
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/data/CassandraUsersRepositoryModule.java]: It doesn't look related to backends

File: server/container/spring/src/main/java/org/apache/james/container/spring/mailbox/SpringUsernameValidator.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ ****************************************************************/
+
+package org.apache.james.container.spring.mailbox;
+
+import org.apache.james.core.User;
+import org.apache.james.mailbox.model.MailboxConstants;
+import org.apache.james.user.api.UsersRepositoryException;
+import org.apache.james.user.lib.UsernameValidator;
+
+public class SpringUsernameValidator implements UsernameValidator {
```
Comment (mbaechler) [inline server/container/spring/src/main/java/org/apache/james/container/spring/mailbox/SpringUsernameValidator.java]: nothing here related to spring so it probably doesn't deserve this name (and no, I'm not an NaaS)

File: server/container/spring/src/main/resources/META-INF/org/apache/james/spring-server.xml [hunk 5]
```diff
@@ -316,4 +316,7 @@
 
     <bean id="jspfLogger" class="org.apache.james.smtpserver.fastfail.SPFHandler.SPFLogger"/>
 
+    <!-- To avoid a circular depandancy with Spring-->
```
Comment (mbaechler) [inline server/container/spring/src/main/resources/META-INF/org/apache/james/spring-server.xml]: dependency

File: server/data/data-library/src/main/java/org/apache/james/user/lib/AbstractUsersRepository.java [hunk 6]
```diff
@@ -20,36 +20,44 @@
...
     /**
      * @see
      * org.apache.james.lifecycle.api.Configurable#configure(org.apache.commons.configuration.HierarchicalConfiguration)
      */
     public void configure(HierarchicalConfiguration configuration) throws ConfigurationException {
-
-        virtualHosting = configuration.getBoolean("enableVirtualHosting", getDefaultVirtualHostingValue());
-        administratorId = Optional.ofNullable(configuration.getString("administratorId"));
-
+        virtualHosting = configuration.getBoolean(ENABLE_VIRTUAL_HOSTING, getDefaultVirtualHostingValue());
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/user/lib/AbstractUsersRepository.java]: please, please, avoid JIT refactoring like this : stash, commit, unstash !

File: server/container/spring/src/main/java/org/apache/james/container/spring/mailbox/SpringUsernameValidator.java (hunk 4 above)
Comment (aduprat) [inline server/container/spring/src/main/java/org/apache/james/container/spring/mailbox/SpringUsernameValidator.java]: What is the difference with `ImapDelegationUsernameValidator` ?

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/data/CassandraUsersRepositoryModule.java [hunk 7]
```diff
@@ -39,6 +41,9 @@
 public class CassandraUsersRepositoryModule extends AbstractModule {
     @Override
     public void configure() {
+        bind(UsernameValidatorAggregator.class).in(Scopes.SINGLETON);
```
Comment (aduprat) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/data/CassandraUsersRepositoryModule.java]: Binding the same implementation in multiple modules -> extract this binding to a common module.

File: server/data/data-library/src/main/java/org/apache/james/user/lib/AbstractUsersRepository.java [hunk 8]
```diff
@@ -68,29 +76,18 @@ public void setDomainList(DomainList domainList) {
...
-                throw new UsersRepositoryException("Given Username contains a @domainpart but virtualhosting support is disabled");
-            }
-        }
+    @Inject
+    public void setAdditionalUsernameValidator(UsernameValidator additionalUsernameValidator) {
+        this.additionalUsernameValidator = additionalUsernameValidator;
+        generateUserNameValidator();
+    }
+
+    private void generateUserNameValidator() {
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/user/lib/AbstractUsersRepository.java]: it doesn't generate, it defines

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/UsersRoutesTest.java [hunk 9]
```diff
@@ -132,6 +132,31 @@ public void postShouldReturnUserErrorWhenEmptyJsonBody() {
...
+                .body("{}")
+            .when()
+                .put("user.name@" + DOMAIN)
+            .then()
+                .statusCode(400);
+
+            // No user created
+            List<Map<String, String>> users =
+                when()
+                    .get()
```
Comment (aduprat) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/UsersRoutesTest.java]: indent

File: server/container/cli-integration/src/test/java/org/apache/james/cli/DataCommandsIntegrationTest.java [hunk 10]
```diff
@@ -110,6 +111,17 @@ public void addUserShouldWork() throws Exception {
...
     }
 
+    @Test
+    public void addUserShouldNotAddUserWithIMAPSeparator() throws Exception {
+        dataProbe.addDomain(DOMAIN);
+
+        assertThatThrownBy(() ->
+            ServerCmd.doMain(new String[] {"-h", "127.0.0.1", "-p", "9999", "ADDUSER", "user.name@domain", PASSWORD}))
+            .hasMessage("User name local part can not contain '.'");
+
```
Comment (aduprat) [inline server/container/cli-integration/src/test/java/org/apache/james/cli/DataCommandsIntegrationTest.java]: This assert is not needed.

File: server/data/data-library/src/main/java/org/apache/james/user/lib/AbstractUsersRepository.java [hunk 11]
```diff
@@ -20,36 +20,44 @@
...
 
 public abstract class AbstractUsersRepository implements UsersRepository, Configurable {
-
-    private DomainList domainList;
+    public static final String ENABLE_VIRTUAL_HOSTING = "enableVirtualHosting";
+    public static final String ADMINISTRATOR_ID = "administratorId";
     private boolean virtualHosting;
     private Optional<String> administratorId;
+    private DomainList domainList;
+    protected UsernameValidator usernameValidator = new UsernameValidatorAggregator(ImmutableSet.of());
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/user/lib/AbstractUsersRepository.java]: you increased complexity of the code by sharing more things in the hierarchy. I would rather extract things to another class to keep responsibility as small as possible here.

File: server/data/data-jdbc/src/main/java/org/apache/james/user/jdbc/DefaultUsersJdbcRepository.java [hunk 12]
```diff
@@ -71,7 +71,7 @@ public void addUser(String username, String password) throws UsersRepositoryExce
         if (contains(username)) {
             throw new AlreadyExistInUsersRepositoryException("User " + username + " already exist");
         }
-        isValidUsername(username);
+        usernameValidator.validate(username);
```
Comment (mbaechler) [inline server/data/data-jdbc/src/main/java/org/apache/james/user/jdbc/DefaultUsersJdbcRepository.java]: previous call looked perfect, why this change ?

File: server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/CassandraUsersRepository.java [hunk 13]
```diff
@@ -186,12 +186,6 @@ public int countUsers() throws UsersRepositoryException {
             .iterator();
     }
 
-    @Override
```
Comment (mbaechler) [inline server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/CassandraUsersRepository.java]: yet another case of "I removed that code in the middle of something else". What's wrong with this code ?

File: server/data/data-library/src/main/java/org/apache/james/user/lib/VirtualHostingNameValidator.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+    public VirtualHostingNameValidator(boolean supportsVirtualHosting, DomainList domainList) {
+        this.supportsVirtualHosting = supportsVirtualHosting;
+        this.domainList = domainList;
+    }
+
+    public void validate(String username) throws UsersRepositoryException {
+        Optional<String> domain = User.fromUsername(username).getDomainPart();
+
+        if (supportsVirtualHosting) {
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/user/lib/VirtualHostingNameValidator.java]: and so it validates non-virtualHosting usernames ? it's not what the name suggest

File: server/protocols/protocols-imap4/src/main/java/org/apache/james/imapserver/mailbox/ImapDelegationUsernameValidator.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+package org.apache.james.imapserver.mailbox;
+
+import javax.inject.Inject;
+
+import org.apache.james.core.User;
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.user.api.UsersRepositoryException;
+import org.apache.james.user.lib.UsernameValidator;
+
+public class ImapDelegationUsernameValidator implements UsernameValidator {
```
Comment (mbaechler) [inline server/protocols/protocols-imap4/src/main/java/org/apache/james/imapserver/mailbox/ImapDelegationUsernameValidator.java]: It's impossible to understand how forbidding the MailboxManager delimiter is about ImapDelegation.

**Comment by aduprat** [comment]
Alternative solution:
https://github.com/linagora/james-project/pull/1108

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/pull/1108
